Track popup to cursor [ready]

peterqliu opened the pull request:
<!-- Thanks for the PR! Feel free to add or remove items from the checklist. -->

<!-- If your PR affects documentation relevant to the currently released version, please use `mb-pages` as the base branch. See https://github.com/mapbox/mapbox-gl-js/blob/master/docs/README.md#committing-and-publishing-documentation -->
ticket: https://github.com/mapbox/mapbox-gl-js/issues/7777

Adds a method `.trackPointer()` to the Popup API that repositions the popup to the mouse cursor on every mousemove event, and hides/shows the popup when cursor leaves/enters the map. Mutually overwritable with `.setLngLat`.

### Design

Opted not to extend setLngLat, as it doesn't quite fit those words. Also opted not to extend the popup `options` object, for cases when users need to change behavior after adding to map.

 ## Launch Checklist

 - [x] briefly describe the changes in this PR
 - [x] write tests for all new functionality
 - [x] document any changes to public APIs
 - [x] manually test the debug page
 - [x] tagged `@mapbox/studio` and/or `@mapbox/maps-design` if this PR includes style spec changes

ryanhamley commented:
> Touchscreen behavior: show popup only on touchmove? Not show at all? Console warning when invoking .trackCursor() on touch devices?

I'm trying to imagine a use case for this on mobile. It seems like it would be a very awkward UX as compared to simply touching on each new point to see a new popup. On the other hand, having a desktop only API method seems odd. Do we have other methods that don't work on mobile? If we do enable this on `touchmove`, maybe it would be better to call the method a more generic term like `trackPointer`.

> Close button: won't be clickable when popup shadows mouse. Should we hide? Does that affect any corner cases?

This is a fun consequence. Maybe there needs to be a separate action to close the popup in this case. Would clicking the map to toggle the popup be too weird?

peterqliu commented:
#### Mobile handling

In lieu of hover, I thought about adding cursor-tracked popups to a tap on the map. The interaction pattern for mobile is unfortunately quite complex: `click` is the event that typically corresponds to this, but is nondeterministic. On Webkit browsers where the user has also implemented a `hover` listener, `click` is invoked only on the *second* tap, and not a double-tap at that. Because of this unpredictability, I've removed touchscreen support entirely for the time being.

#### Other changes

- renamed `trackCursor` to `trackPointer`, to reflect the mouse-specific compatibility
- move the show/hide logic to CSS, greatly simplifying the event handler logic
- opted to keep the closeButton by default, as is the current Popup default. The change is easy enough on the user's side

peterqliu commented:
@KravMaguy that is correct, you will lose that functionality if you use a cursor-tracked popup on mobile. In these contexts, most apps display the information in a panel separate from the map: it's a much better practice given the limited screen space, and preserves visibility/interactivity of the map itself.

some examples ⬇️ 

<details>

![img_3115](https://user-images.githubusercontent.com/4551508/51873687-0662cd80-2313-11e9-960e-c9a5ce0b89df.jpeg)
![img_3117](https://user-images.githubusercontent.com/4551508/51873691-09f65480-2313-11e9-81f3-a2fe87db97dc.jpeg)
![img_3116](https://user-images.githubusercontent.com/4551508/51873698-0d89db80-2313-11e9-8a51-c24e89176202.jpeg)

</details>

peterqliu commented:
@KravMaguy you can still `unproject`/`setLngLat` the popup on each touch event, to move the popup there

ryanhamley commented on src/ui/popup.js:
```diff
@@ -95,13 +96,18 @@ export default class Popup extends Evented {
+        if (this._trackPointer) {
+            this._map.on('mousemove', (e) => { this._update(e.point); });
+            this._container.classList.add('mapboxgl-popup-track-pointer');
+        } else this._map.on('move', this._update);
```
I'd wrap the else block in brackets

ryanhamley commented:
> Agree that this is weird, but the specific solution seems like it would vary by user/use case/preference. I'm inclined to let people modify CSS as appropriate.

👍 I do think it makes sense to change the cursor from a grabber to a pointer when in `trackPointer` mode. The idea of this feature is to point at various locations on the map and get information about them, so the pointer feels semantically valid as well as it visually covering less of the popup.

peterqliu commented:
- opted to hide popup when dragging the map -- tracking popup on drag made it harder to see the map when repositioning it takes priority

- cursor is now `pointer` when tracking popup

peterqliu commented:
>If you move the mouse quickly enough, you can end up inside the popup. I'm not entirely sure what the answer here could be, if anything.

Also removed interactivity with pointer-tracked popups, to avoid getting trapped in it 👍

Plantain commented:
Is this likely to make it into a release in the near future? We'd like to use this feature rather than the workarounds in #7777

peterqliu commented:
@Plantain this branch is ready to the extent that I can tell -- looping in some reviewers to see if we can get this across the line.

ryanhamley commented on src/ui/popup.js:
```diff
@@ -283,18 +322,18 @@ export default class Popup extends Evented {
-    _update() {
-        if (!this._map || !this._lngLat || !this._content) { return; }
+    _update(cursor: PointLike) {
+        if (!this._map || !this._lngLat && !this._trackPointer || !this._content) { return; }
```
for clarity's sake, you may want to explicitly group the middle term here: `!this._map || (!this._lngLat && !this._trackPointer) || !this._content`

ryanhamley commented:
This is too late to be in the next release of GL JS which comes out this week, but it should be in May's release

peterqliu commented:
clarified a bit of the code per https://github.com/mapbox/mapbox-gl-js/pull/7786#issuecomment-479154849 and squashed as much as I can (not sure if I can squash past a previous pull from master?)

ryanhamley commented:
You don't have to squash ahead of time @peterqliu. Once all the checks have passed, just pressing the `Squash and merge` button will take care of everything.
